include/nbl/builtin/hlsl/bxdf/transmission/ggx.hlsl

```diff
@@ -0,0 +1,422 @@
+    {
+        return create(params.orientedEta, params.ax, params.ay);
+    }
+
+    query_type createQuery(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction)
+    {
+        query_type query;
+        ndf_type ggx_ndf = __base.ndf;
+        query.devsh_v = ggx_ndf.devsh_part(interaction.getTdotV2(), interaction.getBdotV2(), interaction.getNdotV2());
+        query.devsh_l = ggx_ndf.devsh_part(_sample.getTdotL2(), _sample.getBdotL2(), _sample.getNdotL2());
+        return query;
+    }
+
+    spectral_type eval(NBL_CONST_REF_ARG(query_type) query, NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, NBL_CONST_REF_ARG(anisocache_type) cache)
+    {
+        struct SGGXG2XQuery
+        {
+            using scalar_type = scalar_type;
+
+            scalar_type getDevshV() NBL_CONST_MEMBER_FUNC { return devsh_v; }
+            scalar_type getDevshL() NBL_CONST_MEMBER_FUNC { return devsh_l; }
+            BxDFClampMode getClampMode() NBL_CONST_MEMBER_FUNC { return _clamp; }
+
+            scalar_type devsh_v;
+            scalar_type devsh_l;
+            BxDFClampMode _clamp;
+        };
+
+        SGGXG2XQuery g2_query;
+        g2_query.devsh_v = query.getDevshV();
+        g2_query.devsh_l = query.getDevshL();
+        g2_query._clamp = _clamp;
+
+        fresnel::OrientedEtas<monochrome_type> orientedEta = __base.fresnel.orientedEta;
+        measure_transform_type dualMeasure = __base.template __DG<SGGXG2XQuery>(g2_query, _sample, interaction, cache);
+        dualMeasure.absNdotL = _sample.getNdotL(_clamp);
+        dualMeasure.orientedEta = orientedEta.value[0];
+        scalar_type DG = dualMeasure.getProjectedLightMeasure();
+
+        return hlsl::promote<spectral_type>(__base.fresnel(hlsl::abs(cache.getVdotH()))[0]) * DG;
+    }
+
+    sample_type __generate_wo_clamps(const vector3_type localV, const vector3_type H, NBL_CONST_REF_ARG(matrix3x3_type) m, NBL_REF_ARG(vector3_type) u, NBL_CONST_REF_ARG(fresnel::OrientedEtaRcps<monochrome_type>) rcpEta, NBL_REF_ARG(anisocache_type) cache)
+    {
+        const scalar_type localVdotH = nbl::hlsl::dot<vector3_type>(localV,H);
+        const scalar_type reflectance = __base.fresnel(hlsl::abs(cache.getVdotH()))[0];
+        
+        scalar_type rcpChoiceProb;
+        bool transmitted = math::partitionRandVariable(reflectance, u.z, rcpChoiceProb);
+
+        cache = anisocache_type::createForReflection(localV, H);
+
+        const scalar_type VdotH = cache.iso_cache.getVdotH();
+        Refract<scalar_type> r = Refract<scalar_type>::create(localV, H);
+        cache.iso_cache.LdotH = hlsl::mix(VdotH, r.getNdotT(rcpEta.value2[0]), transmitted);
+        ray_dir_info_type localL;
+        bxdf::ReflectRefract<scalar_type> rr;
+        rr.refract = r;
+        localL = localL.reflectRefract(rr, transmitted, rcpEta.value[0]);
+
+        return sample_type::createFromTangentSpace(localL, m);
```

Comment:
btw when this moves to Cook Torrance, you need to make sure that you somehow "fail" the generation if the following happens:
1. you choose reflect but resulting `localL.z` has different sign to `localV.z`
2. you choose refract but resulting `local.z` has same sign as `localV.z`

This is simply to avoid using samples which have invalid paths.

See https://github.com/Devsh-Graphics-Programming/Nabla/pull/899#discussion_r2244474335

I think the returned aniso cache can be queried for invalidity, or the sample somehow.